rfcs/0042-networking-refactor.md

```diff
@@ -0,0 +1,334 @@
+
+Below is a proposed set of data streams we would setup for the helper IPC protocol. Keep in mind that some of these pipes cannot be setup without helper message type awareness, and thus will not be available separately until we perfrom a hard-fork update.
+
+- stdin (used only for initialization message, then closed)
+- stdout (used only for helper logging)
+- stderr (used only for helper logging)
+- stats\_in (publishes helper stats to daemon on an interval)
+- block\_gossip\_in (incoming block gossip messages)
+- mempool\_gossip\_in (other incoming gossip messages, related to mempool state)
+- response\_in (incoming RPC responses)
+- request\_in (incoming RPC requests)
+- validation\_out (all validations except request validations, which are bundled with responses)
+- data\_out (responses, broadcasts)
+
+The rough priorities for reading the incoming pipes from the daemon process would be:
+
+- block\_gossip\_in
+- response\_in
+- request\_in
+- mempool\_gossip\_in
+
+NOTE: It is critical in the implementation of this that the prioritization scheme we choose here does not allow the mempool gossip pipe to be starved. The main thing to keep in mind to avoid this is to ensure that we do not over weight reading the incoming requests, so that another node on the network cannot delay (or potentially censor) txns and snarks we are receiving over gossip. One approach towards this could be to limit the parallelism per pipe while keeping the maximum parallel messages we handle from IPC high enough such that we can always schedule new mempool gossip jobs even when there are a lot of requests.
+
+CONSIDER: Is it important that IPC messages include timestamps so that the daemon and helper processes can perform staleness checks as they read messages? For example: if we haven't read a mempool gossip in a while, and read one, discovering that the first message on the pipe is rather old, should we further prioritize this pipe for a bit until we catchup? A potential risk of this system is that it would be hard to guarantee that none of the data streams aren't succeptible to starvation attacks.
+
+#### Serialization Format
+
+The new serialization format will be [Cap'N Proto](https://capnproto.org/). There are already [Go](https://github.com/capnproto/go-capnproto2) and [OCaml](https://github.com/capnproto/capnp-ocaml) libraries for the Cap'N Proto serialization format, which generate code for each language based on a common schema definition of the protocol. Using Cap'N Proto instead of JSON will allow us to embed raw binary data in our IPC messages, which will avoid the rather costly and constant base64 encoding/decoding we currently do for all binary data we transfer between the processes. It's possible to keep some pipes in JSON if preferable, but within the current plan, all messages would be converted over to Cap'N Proto to avoid having to support tooling for keeping both serialization formats in sync between the processes.
```

Comment:
I've poked around a bit, Cap'N Proto looks pretty nice!

The OCaml implementation seems to compile down to efficient machine code for integers (at least, for ints < 32 bits; beyond that we use the OCaml heap at some additional cost). However, if we want to get the best efficiency, we may want to tweak their handling of binary data; presumably this will be the largest part of the payload for most messages, and avoiding redundant copying should give us some benefits.

Currently, their default implementation uses `Bytes.t` as the backing type, and uses a `string` to represent the 'data' kind. We would save 2 copies (1 to create the string from the data field, 1 to convert it to a `Bigstring.t` for bin_prot) if we
* used `Bigstring.t` as the backing type (as simple as reimplementing [the interface that `BytesStorage` uses](https://github.com/capnproto/capnp-ocaml/blob/master/src/runtime/bytesStorage.ml) for it)
* emitted a slice of the backing type for bytes (by changing [here](https://github.com/capnproto/capnp-ocaml/blob/dda3d811aa7734110d7af051465011f1f823ffb9/src/runtime/readerInc.ml#L418) and a few places in the codegen).

(Obviously, still, using the drop-in replacement is better than base64-encoding everything :grin:)

Reply:
Good catch. Yeah, seems like we should fork that and remove the unnecessary message copies. I will make note of that in the RFC.

Reply:
I like this idea -- but could we also try to upstream the changes too?